webhook/pr_review.go

```diff
@@ -0,0 +1,384 @@
+			time.Sleep(time.Duration(sleepTimeSecond) * time.Second)
+			builder.Reset()
+		}
+	}
+}
+
+func getChatGPTComment(patch string) (string, error) {
+	client, err := openai.New(
+		openai.WithToken(viper.GetString("openai.api_key")),
+		openai.WithModel(viper.GetString("openai.model")),
+		openai.WithOrgID(viper.GetString("openai.org_id")),
+		openai.WithProxyURL(viper.GetString("openai.proxy")),
+		openai.WithSocksURL(viper.GetString("openai.socks")),
+		openai.WithBaseURL(viper.GetString("openai.base_url")),
+		openai.WithTimeout(viper.GetDuration("openai.timeout")),
+		openai.WithMaxTokens(viper.GetInt("openai.max_tokens")),
+		openai.WithTemperature(float32(viper.GetFloat64("openai.temperature"))),
+	)
+	if err != nil {
+		log.Println("Error creating OpenAI client:", err)
+		return "", err
+	}
+
+	out, err := util.GetTemplateByString(
+		prompt.CodeReviewTemplate,
+		util.Data{
+			"file_diffs": patch,
+		},
+	)
+	if err != nil {
+		log.Println("Error creating prompt:", err)
+		return "", err
+	}
+
+	ctx := context.Background()
+	resp, err := client.Completion(ctx, out)
+	if err != nil {
+		log.Println("Error completing prompt:", err)
+		return "", err
+	}
+	summarizeMessage := resp.Content
+
+	log.Println("PromptTokens: " + strconv.Itoa(resp.Usage.PromptTokens) +
+		", CompletionTokens: " + strconv.Itoa(resp.Usage.CompletionTokens) +
+		", TotalTokens: " + strconv.Itoa(resp.Usage.TotalTokens))
+
+	// Output core review summary
+	log.Println("================Review Summary====================")
+	log.Println("\n" + strings.TrimSpace(summarizeMessage) + "\n\n")
+	log.Println("==================================================")
+
+	return summarizeMessage, nil
+}
+
+func getLastLinePosition(patch string) int {
+	lines := strings.Split(patch, "\n")
+	position := 0
+	for _, line := range lines {
+		if strings.HasPrefix(line, "@@") {
+			continue
+		}
+		position++
+	}
+	return position
+}
+
+func submitReviewComment(owner, repo string, prNumber int, commitID, filePath, comment string, position int) error {
+	client := &http.Client{}
+
+	commentRequest := CommentRequest{
+		Body:     comment,
+		Path:     filePath,
+		Position: position,
+		CommitID: commitID,
+	}
+
+	jsonData, err := json.Marshal(commentRequest)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf(submitCommentsURL, apiBaseurl, owner, repo, prNumber), bytes.NewBuffer(jsonData))
+	if err != nil {
+		log.Println("Error creating request:", err)
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "token "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Error submitting review comment:", err)
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusCreated {
+		log.Println("Review comment submitted successfully")
+	} else {
+		return fmt.Errorf("error: %d", resp.StatusCode)
+	}
+
+	return nil
+}
```

Comment:
Overall, the code patch looks fine and there are no immediate bug risks or security vulnerabilities. 

However, there are some improvement suggestions: 

1. Error handling: The error handling could be improved by providing more descriptive error messages or logging the errors in a more detailed manner. For example, instead of just logging "Error creating request" or "Error submitting review comment", the specific error message returned by the API could be included in the log message. 

2. Magic strings: The string literals like "Content-Type" and "Authorization" could be defined as constants or variables to avoid using magic strings in the code. 

3. Code organization: The code could be organized into separate functions or methods for creating the request, setting headers, sending the request, and handling the response. This would make the code more modular and easier to read and maintain. 

4. Input validation: It would be a good practice to validate the input parameters like the token before using them in the code. 

5. Error handling in defer: The defer statement should be used with caution when handling errors as it may lead to silent failures. In this case, it would be better to handle the error returned by resp.Body.Close() instead of relying on the defer statement.